### PR TITLE
TSV-402: error logging (with stacktrace) on backend failures

### PR DESCRIPTION
**Pull request by scottkwalker**

### CODE REVIEW

File: src/test/scala/uk/gov/hmrc/play/microservice/bootstrap/JsonErrorHandlingSpec.scala [hunk 1]
```diff
@@ -51,6 +55,35 @@ class JsonErrorHandlingSpec extends WordSpecLike with Matchers with ScalaFutures
...
+          message should include regex s"! @[a-zA-Z0-9]+ - Internal server error, for \\($method\\) \\[$uri\\] ->".r
+        }
+      }
+    }
+
+    "log an error" in {
+      withCaptureOfLoggingFrom(Logger) { logEvents =>
+        val method = "some-method"
+        val uri = "some-uri"
+        when(requestHeader.method).thenReturn(method)
```
Comment (paoloboni) [inline src/test/scala/uk/gov/hmrc/play/microservice/bootstrap/JsonErrorHandlingSpec.scala]: verify statements on this mock are missing

File: src/test/scala/uk/gov/hmrc/play/microservice/bootstrap/JsonErrorHandlingSpec.scala [hunk 2]
```diff
@@ -51,6 +55,35 @@ class JsonErrorHandlingSpec extends WordSpecLike with Matchers with ScalaFutures
...
+        when(requestHeader.uri).thenReturn(uri)
+
+        jsh.onError(requestHeader, new Exception("any application exception")).futureValue
+
+        eventually {
+          logEvents.filter(_.getLevel == Level.ERROR).loneElement.getMessage should include(s"! Internal server error, for ($method) [$uri] ->")
```
Comment (paoloboni) [inline src/test/scala/uk/gov/hmrc/play/microservice/bootstrap/JsonErrorHandlingSpec.scala]: With this statement you probably want to make sure the _only_ error log recorded by the test appender is the one with:
- the expected level (ERROR)
- the expected message
So I don't see the point of filtering logEvents.
Also, the assertion on the log message could be stronger (shouldBe).

File: src/main/scala/uk/gov/hmrc/play/microservice/bootstrap/JsonErrorHandling.scala [hunk 3]
```diff
@@ -32,18 +33,37 @@ trait JsonErrorHandling extends GlobalSettings {
...
+    logError(request, ex)
+    Future.successful(resolveError(request, ex))
+  }
 
-        case e: Upstream4xxResponse => (e.reportAs, e.getMessage)
-        case e: Upstream5xxResponse => (e.reportAs, e.getMessage)
+  private def logError(request: RequestHeader, ex: Throwable): Unit = {
+    def formatPlayExceptionId: Throwable => String = {
+      case p: PlayException => "@" + p.id + " - "
+      case _ => ""
```
Comment (AndyNewton) [inline src/main/scala/uk/gov/hmrc/play/microservice/bootstrap/JsonErrorHandling.scala]: for 'case _' I'd much rather we return some representation of the Throwable and not ""

File: src/main/scala/uk/gov/hmrc/play/microservice/bootstrap/JsonErrorHandling.scala [hunk 4]
```diff
@@ -32,18 +32,24 @@ trait JsonErrorHandling extends GlobalSettings {
...
-
-        case e: Upstream4xxResponse => (e.reportAs, e.getMessage)
-        case e: Upstream5xxResponse => (e.reportAs, e.getMessage)
+    logError(request, ex)
+    Future.successful(resolveError(ex))
+  }
 
-        case e: Throwable => (INTERNAL_SERVER_ERROR, e.getMessage)
-      }
+  private def logError(request: RequestHeader, ex: Throwable): Unit = {
```
Comment (paoloboni) [inline src/main/scala/uk/gov/hmrc/play/microservice/bootstrap/JsonErrorHandling.scala]: Now that exception handling has been removed (as it was unnecessary), can this `logError` be inlined in `onError`? It looks like there's a single statement here.

File: src/test/scala/uk/gov/hmrc/play/microservice/bootstrap/JsonErrorHandlingSpec.scala [hunk 5]
```diff
@@ -16,41 +16,64 @@
...
-  val requestHeader = mock[RequestHeader]
+class JsonErrorHandlingSpec extends UnitSpec with ScalaFutures with MockitoSugar with LogCapturing with LoneElement with Eventually {
 
   "error handling in onError function" should {
 
-    "convert a NotFoundException to NotFound response" in {
+    "convert a NotFoundException to NotFound response" in new Setup {
       val resultF = jsh.onError(requestHeader, new NotFoundException("test")).futureValue
       resultF.header.status shouldBe 404
     }
```
Comment (AndyNewton) [inline src/test/scala/uk/gov/hmrc/play/microservice/bootstrap/JsonErrorHandlingSpec.scala]: @scottkwalker Whilst you are here, what do you think of adding another check, to each of these existing cases, for the json generated, i.e. for the result of the `Json.toJson(errorResponse)`?  Let me know if that doesn't make sense - I'm only looking at the diffs on the PR.  Thanks.

**Pull request merged**